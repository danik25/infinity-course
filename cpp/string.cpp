/****************************************************************
*                              String
* 						=====================
* File Name: string.cpp
* Related files: string.hpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#include <cstring> /* strlen */
#include <iostream> /* cout */
#include "string.hpp"

namespace hrd9
{
	/* constructor */
	String::String(const char* str)
	{
		size_t len = strlen(str);
		m_str = new char [len + 1];
		strcpy(m_str, str);
	}

	/* distructor */
	String::~String()
	{
		delete [](m_str);
	}

	/* copy constructor */
	String::String(const String& other) : m_str(init_new_str(other.m_str))
	{
		;
	}

	/* copy assignment operator */
	String& String :: operator=(const String& other)
	{
		delete[] m_str;
		m_str = init_new_str(other.m_str);

		return(*this);
	}

/*===================== non member functions =====================*/

/*===================== member functions =====================*/

	String& String::operator+=(const String& other)
	{
		size_t total_len = strlen(m_str) + strlen(other.m_str);

		String tmp = m_str;
		delete [] m_str;

		m_str = new char [total_len + 1];
		char * runner = m_str;

		strcpy(runner, tmp.m_str);
		strcat(runner, other.m_str);

		return(*this);
	}

	size_t String::length() const
	{
		return(strlen(m_str));
	}

	const char& String::operator[](size_t index) const
	{
		return(m_str[index]);
	}

	char& String::operator[](size_t index)
	{
		return(m_str[index]);
	}

	const char* String::to_c_str() const
	{
		return(m_str);
	}

/*===================== friend functions =====================*/

	const String operator+(const String& lhs, const String& rhs)
	{
		String ans = lhs;
		ans += rhs;

		return(ans);
	}

	bool operator==(const String& lhs, const String& rhs)
	{
		return(!strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator!=(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator>(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator<(const String& lhs, const String& rhs)
	{
		return(strcmp(rhs.m_str, lhs.m_str));
	}

	std::ostream& operator<<(std::ostream& os, const String& str)
	{
		os << str.m_str;
		return(os);
	}

/*===================== assistance functions =====================*/

	char* String::init_new_str(const char *str)
	{
		size_t len = strlen(str);
		char *new_str = new char [len + 1];
		strcpy(new_str, str);

		return(new_str);
	}
}