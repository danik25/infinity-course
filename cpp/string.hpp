/****************************************************************
*                              String
* 						=====================
* File Name: string.hpp
* Related files: string.cpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#ifndef ILHRD_STRING_HPP
#define ILHRD_STRING_HPP

#include <iosfwd> //ostream forward declarations

namespace hrd9
{

class String;
//non member functions:
const String operator+(const String& lhs, const String& rhs);
bool operator==(const String& lhs, const String& rhs);
bool operator!=(const String& lhs, const String& rhs);
bool operator>(const String& lhs, const String& rhs);
bool operator<(const String& lhs, const String& rhs);
std::ostream& operator<<(std::ostream& os, const String& str);
std::istream& operator>>(std::istream& is, String& str);

class String
{

public:
	// explicitly non explicit constructor
	String(const char* str = "");
	String(const String& other);
	String& operator=(const String& other);
	~String();

	String& operator+=(const String& other);
	size_t length() const;

	const char& operator[](size_t index) const; //read string[]
	char& operator[](size_t index);             //string[i] assignment

	const char* to_c_str() const;

private:
	friend bool operator==(const String& lhs, const String& rhs);
	friend bool operator!=(const String& lhs, const String& rhs);
	friend bool operator>(const String& lhs, const String& rhs);
	friend bool operator<(const String& lhs, const String& rhs);
	friend std::ostream& operator<<(std::ostream& os, const String& str);
	friend std::istream& operator>>(std::istream& is, String& str);

	char* init_new_str(const char *str);

	char* m_str;
};
} //namespace hrd9




#endif /* ILHRD_STRING_HPP */