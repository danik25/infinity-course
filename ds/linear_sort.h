/*******************************************************************************
*                Data Structures - Linear Sorting   
* file name: linear_sort.h                                              
* Related files: linear_sort.c linear_sort_test.c                               
* #Version: V 2.1                                                  
* writer: Dani Benjamin                                                         
* reviewer: Lev Shubert                                                 
* last update: 14.5.2019                                               
*******************************************************************************/ 

#ifndef _LINEAR_SORT_H_
#define _LINEAR_SORT_H_

#include <stddef.h>  /* size_t */

/* Implementation of two linear non-comparison sorting algorithms:
 * COUNTING SORT and RADIX SORT.
 */  


/* The signature of the function provided by the user with the sorting rule.
 * the function get key and convert it to number.
 */
typedef size_t (*key_to_num_t)(const void *key, const void *arg);



/* COUNTING SORT
 *
 * Description:
 *      An algorithm for sorting a collection of objects 
 *      according to keys that are non-negativ integers.
 *      It operates by counting the number of objects 
 *      that have each distinct key value, 
 *      and using arithmetic on those counts to determine 
 *      the positions of each key value in the result sequence.  
 *      
 *      This implementation preserves stability.            
 *
 * Params: 
 *      arr - containes the elements to be sorted.
 *      len - number of elements in the array.
 *      element_size - element size in bytes.
 *      KeyToNum - from-key-to-number conversion function.
 *      arg - argument for KeyToNum function.
 *      range - range of values in the array.
 *        
 * Return Value: 
 *      1 - fail
 *      0 - success
 *  
 * Time complexity:
 *      O(n + k)      
 */
int CountingSort(void *arr, size_t len, size_t element_size,
                 key_to_num_t KeyToNum, const void *arg, size_t range);

/*
 * RADIX SORT
 *
 * Description:
 *      a non-comparative integer sorting algorithm 
 *      that sorts data with integer keys by grouping keys 
 *      by the individual digits which share the same 
 *      significant position and value.            
 *
 * Params: 
 *      arr - containes the elements to be sorted.
 *      len - number of elements in the array.
 *      element_size - element size in bytes.
 *      KeyToNum - from-key-to-number conversion function.
 *      arg - argument for KeyToNum function.
 *      max_digits_num - maximum digits of element in the array.
 *        
 * Return Value: 
 *      1 - fail
 *      0 - success
 *  
 * Time complexity:
 *      O(n + k)      
 */
int RadixSort(void *arr, size_t len, size_t element_size,
                 key_to_num_t KeyToNum, const void *arg, size_t max_digits_num);


#endif /* _LINEAR_SORT_H_ */