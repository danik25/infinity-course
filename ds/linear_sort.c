/*******************************************************************************
*                Data Structures - Linear Sorting   
* file name: linear_sort.c                                             
* Related files: linear_test.c linear_sort.c                               
* #Version: V 2.1                                                  
* writer: Dani Benjamin                                                         
* reviewer: Lev Shubert                                                     
* last update: 14.5.2019                                               
*******************************************************************************/
#include <stdlib.h> /* malloc */
#include <assert.h> /* assert */
#include <string.h> /* memcpy */
#include "linear_sort.h"

#define UNUSED(x) (void)(x)
static size_t Key(const void *key, const void *arg);

typedef struct assistance_struct
{
    size_t index;
    key_to_num_t user_func;
    const void* user_arg;
}assistance_struct_t;

int CountingSort(void *arr, size_t len, size_t element_size,
                 key_to_num_t KeyToNum, const void *arg, size_t range)
{
    size_t *index = NULL;
    char * tmp = NULL;
    char * runner = NULL;
    size_t i = 0;

    assert(arr);
    assert(KeyToNum);

    runner = arr;
    index = (size_t*)calloc(range, sizeof(size_t));

    if(NULL == index)
    {
        return (1);
    }

    tmp = (char*)malloc(len * element_size);

    if(NULL == tmp)
    {
        free(index);
        index = NULL;
        return (1);
    }

    /* filling the index lut */
    for(i = 0; i < len; ++i, runner += element_size)
    {
        *(index + KeyToNum(runner, arg)) += 1;
    }

    /* adjusting the index lut */
    i = 1;
    for(i = 1; i < range; ++i)
    {
        *(index + i) = *(index + i) + *(index + i - 1);
    }

    /* filling the temporary array */
    i = 0;
    runner = (char*)arr + (len - 1) * element_size;
    for(i = 0; i < len; ++i, runner -= element_size)
    {
        memcpy(tmp + (index[KeyToNum(runner, arg)] - 1) * element_size, 
                                                runner, element_size);
        index[KeyToNum(runner, arg)] -= 1;
    }

    memcpy(arr, tmp, len * element_size);

    free(tmp);
    tmp = NULL;
    free(index);
    index = NULL;

    return (0);
}

int RadixSort(void *arr, size_t len, size_t element_size,
                 key_to_num_t KeyToNum, const void *arg, size_t max_digits_num)
{
    size_t i = 0;
    assistance_struct_t argument = {0 ,0 , 0}; 

    assert(arr);
    assert(KeyToNum);

    argument.index = 0;
    argument.user_func = KeyToNum;
    argument.user_arg = arg;
    
    for(i = 0; i < max_digits_num; ++i)
    {
        if(1 == CountingSort(arr, len, element_size, Key, &argument, 10))
        {
            return(1);
        }
        ++argument.index;     
    }    
    return(0);
}

static size_t Key(const void *key, const void *arg)
{
    assistance_struct_t *func_struct = NULL;
    size_t num = 0;
    size_t i = 0;

    assert(key);
    assert(arg);

    func_struct = (assistance_struct_t *)arg;
    num = func_struct->user_func(key, func_struct->user_arg);

    for(i = 0; i < func_struct->index; ++i)
    {
        num /= 10;
    }

    return((size_t)((num%10)));
}