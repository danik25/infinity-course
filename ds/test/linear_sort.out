/home/xyz/git/dani-benjamin/ds/linear_sort.out